From f5c5669935eb1ea9eb05f8cd9684b4bcb23b9454 Mon Sep 17 00:00:00 2001
From: pjgowtham <pjgowtham@gmail.com>
Date: Thu, 22 May 2025 12:46:34 +0530
Subject: [PATCH 2/4] sdm: Operate FINGERPRINT_MASK for oplus optical UDFPS

Even though oplus calls the connector property hbm_enable, its
function is not just limited to operating HBM. It's also responsible
for other aspects of UDFPS operations like that of setFpPress so
that we don't have to handle any of these through IOCTL.

The purpose of this DRM connector property is for the userspace to
let the kernel display and fingerprint drivers know which layers are
visible. The same can be achieved by asserting with the layer name
instead of setting a magic constant for the layer ZOrder.

Here's what stock uses:

| Bit index | Decimal | Layer       | Z-order      |
| --------- | ------- | ----------- | ------------ |
| 0         | 1       | **dim**     | `0x41000005` |
| 1         | 2       | **pressed** | `0x41000033` |
| 2         | 4       | **icon**    | `0x41000032` |
| 3         | 8       | **aod**     | `0x41000011` |

Change-Id: I9fcb119bbebf44b2cd13fb465c09cd9d332013ed
---
 Android.bp                          | 11 +++++++++
 sde-drm/drm_property.cpp            |  2 +-
 sdm/libs/core/drm/hw_device_drm.cpp | 35 +++++++++++++++++++++++++++++
 3 files changed, 47 insertions(+), 1 deletion(-)

diff --git a/Android.bp b/Android.bp
index b96f7aee02..63b3fa9461 100644
--- a/Android.bp
+++ b/Android.bp
@@ -8,6 +8,7 @@ soong_config_module_type {
         "llvmsa",
         "gralloc4",
         "udfps",
+        "oplus_udfps",
         "samsung_udfps",
         "default",
         "var3",
@@ -17,6 +18,7 @@ soong_config_module_type {
         "enabled",
         "srcs",
         "header_libs",
+        "include_dirs",
     ],
 }
 
@@ -108,6 +110,15 @@ qtidisplay_cc_defaults {
                 enabled: true,
             },
         },
+        oplus_udfps: {
+            cflags: [
+                "-DCONNECTOR_PROP_UDFPS",
+                "-DOPLUS_FINGERPRINT_MASK",
+            ],
+            include_dirs: [
+                "frameworks/native/services/surfaceflinger/CompositionEngine/include/compositionengine"
+            ],
+        },
         samsung_udfps: {
             cflags: [
                 "-DCONNECTOR_PROP_UDFPS",
diff --git a/sde-drm/drm_property.cpp b/sde-drm/drm_property.cpp
index 1f789f531f..602356bd9c 100644
--- a/sde-drm/drm_property.cpp
+++ b/sde-drm/drm_property.cpp
@@ -181,7 +181,7 @@ DRMProperty DRMPropertyManager::GetPropertyEnum(const std::string &name) const {
   if (name == "dimming_dyn_ctrl") { return DRMProperty::DIMMING_DYN_CTRL; }
   if (name == "dimming_min_bl") { return DRMProperty::DIMMING_MIN_BL; }
 #ifdef CONNECTOR_PROP_UDFPS
-  if (name == "fingerprint_mask") { return DRMProperty::FINGERPRINT_MASK; }
+  if (name == "fingerprint_mask" || name == "hbm_enable") { return DRMProperty::FINGERPRINT_MASK; }
 #endif
 
   return DRMProperty::INVALID;
diff --git a/sdm/libs/core/drm/hw_device_drm.cpp b/sdm/libs/core/drm/hw_device_drm.cpp
index a1eee4fff6..04ae076386 100644
--- a/sdm/libs/core/drm/hw_device_drm.cpp
+++ b/sdm/libs/core/drm/hw_device_drm.cpp
@@ -106,6 +106,23 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "hw_device_drm.h"
 #include "hw_info_interface.h"
 
+#ifdef OPLUS_FINGERPRINT_MASK
+#define BIT(nr) (1UL << (nr))
+enum oplus_ofp_property_value {
+    OPLUS_OFP_PROPERTY_NONE = 0,
+    OPLUS_OFP_PROPERTY_DIM_LAYER = BIT(0),
+    OPLUS_OFP_PROPERTY_FINGERPRESS_LAYER = BIT(1),
+    OPLUS_OFP_PROPERTY_ICON_LAYER = BIT(2),
+    OPLUS_OFP_PROPERTY_AOD_LAYER = BIT(3),
+};
+
+#include <UdfpsExtension.h>
+#ifdef UDFPS_TOUCHED_LAYER_NAME
+#undef UDFPS_TOUCHED_LAYER_NAME
+#define UDFPS_TOUCHED_LAYER_NAME "SurfaceView[UdfpsControllerOverlay]"
+#endif
+#endif
+
 #define __CLASS__ "HWDeviceDRM"
 
 #ifndef DRM_FORMAT_MOD_QCOM_COMPRESSED
@@ -1427,6 +1444,24 @@ void HWDeviceDRM::SetupAtomic(Fence::ScopedRef &scoped_ref, HWLayersInfo *hw_lay
   }
 #endif
 
+#ifdef OPLUS_FINGERPRINT_MASK
+  if (IsPrimaryDisplay()) {
+    uint32_t mask_state = 0;
+    for (uint32_t i = 0; i < hw_layer_count; i++) {
+      const std::string &name = hw_layers_info->hw_layers[i].layer_name;
+      if (name.find(UDFPS_DIM_LAYER_NAME) != std::string::npos) {
+        mask_state |= OPLUS_OFP_PROPERTY_DIM_LAYER;
+      } else if (name.find(UDFPS_TOUCHED_LAYER_NAME) != std::string::npos) {
+        mask_state |= OPLUS_OFP_PROPERTY_FINGERPRESS_LAYER;
+      }
+    }
+    if (current_mask_state_ != mask_state) {
+      current_mask_state_ = mask_state;
+      drm_atomic_intf_->Perform(DRMOps::CONNECTOR_SET_FINGERPRINT_MASK, token_.conn_id, mask_state);
+    }
+  }
+#endif
+
   if (hw_panel_info_.partial_update && update_config) {
     if (IsFullFrameUpdate(*hw_layers_info)) {
       ResetROI();
-- 
2.43.0

